bugfix for gh-2239:  Merging views with global elements

sw96411 opened the pull request:
Resolves #2293 

Only expand Global* elements in a View if the schema contains an entry with that element type in.  This supports stores such as the FederatedStore where the first Store a query gets run against is in some way a proxy of some other store(s) and doesn't have the schema information to do the expansion.

Also, update the logic in updateOperationChain such that it no longer expands  a View with no entities and edges but, e.g., a globalEntities element to include all of the edges.

When reviewing, please note that I've chosen to change updateOperationChain to default scope for testing purposes and because the code didn't feel like it was perfectly encapsulated by the logic of executing a query.  Alternatively, I could change the tests to run against the public execute method (which felt a bit broad) or use reflection to test the method as a private (but I'm not sure how you feel about that 'round these parts).  Lastly, where there's identical logic for Edges and Entities, I've only included a unit for the edges part, but let me know if you want a repeated unit test for the Entities logic

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gchq/Gaffer?pullRequest=2343) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/gchq/Gaffer?pullRequest=2343) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gchq/Gaffer?pullRequest=2343) it.</sub>

sw96411 commented:
It looks like a CI build failed because travis couldn't download an apache commons library from maven - I don't think that's anything to do with this change; is it possible for someone re-run the checks?

d47853 commented on core/graph/src/main/java/uk/gov/gchq/gaffer/graph/Graph.java:
```diff
@@ -350,22 +354,32 @@ public JobDetail executeJob(final Job job, final Context context) throws Operati
     }
 
-    private void updateOperationChainView(final Operations<?> operations) {
+    void updateOperationChainView(final Operations<?> operations) {
```
Is this change needed? If it's just for the test then it would be better to indirectly call it from one of the public methods

sw96411 replied:
Sure thing, this is what I was trying to get at in the PR description - I'll refactor the test to work on the execute method and revert the visibility of the method.

d47853 commented on core/graph/src/test/java/uk/gov/gchq/gaffer/graph/GraphTest.java:
```diff
@@ -166,73 +167,41 @@ public void shouldConstructGraphFromSchemaModules() {
-                        .destination("vertex")
-                        .directed(DIRECTED_EITHER)
-                        .build())
+                .type(TestTypes.PROP_STRING, new TypeDefinition.Builder().clazz(String.class).build())
```
There are a lot of changes to this file, most are whitespace from what I can tell. Please could you revert some of the whitespace changes so it's easier to review?

sw96411 replied:
Sorry about that - was trying to setup my save actions to work with the Gaffer checkstyle and it looks like it's started merging lines (this also seems to have caused the measured test coverage % to drop) - I'll sort that out

d47853 commented on core/graph/src/test/java/uk/gov/gchq/gaffer/graph/GraphTest.java:
```diff
@@ -2418,20 +1970,94 @@ public void shouldNotAddDefaultSecurityHooksIfAlreadyAdded() {
+    }
+
+    @Test
+    public void shouldExpandGlobalEdges() {
```
Might be worth adding some integration tests into the federated store to check the behaviour

sw96411 commented:
@d47853 :  I think I've addressed the points you discussed.  The automated PR checks show the unit tests are failing, but I note they're failing the same way with other open PRs right now, and in a way that doesn't seem to be to do with this change.  Is there something up with the CI build at the moment?  If so, happy to wait for that to get resolved and then re-run the checks here.

d47853 commented:
Yeah it looked like the Flink tests failed which we're looking at as part of #2325. I've re-run the CI

sw96411 commented:
Looks like we've got a little further, and I've added some tests to improve coverage, but it looks like the checks are failing again - I note for the same reasons as #2344 .  Do you think this is something I need to resolve in this PR, @d47853 , or is this something up with the build?  I've run the mvn build executed by the REST CI build locally and it seems fine, and the coverage of my change is higher than the project overall, and I've tried to verify locally with Emma, although I appreciate different coverage tools may use different metrics

d47853 commented:
No I think that's something that should be looked at as part of a separate issue. I've re-run the tests as the failure seems to be intermittent

d47853 commented:
As for coverage, we won't get a full picture until all the tests pass. You can imagine that some modules have slightly different coverages so when one or more reports aren't uploaded, we don't get a full picture

d47853 commented:
That failure now appears to happen pretty consistently. I've raised a new issue for it: #2345. Once a fix for it is added, we can re-run the tests and merge if successful.
